R/txrange.R

```diff
@@ -0,0 +1,282 @@
+    fix="start")[which(as.character(seqnames(RCranges)) %in%
+    names(genomeSequence))])
+  endSeqs = BSgenome::getSeq(genomeSequence,resize(RCranges,10,
+    fix="end")[which(as.character(seqnames(RCranges)) 
+    %in% names(genomeSequence))])
+  #count number of A's in the first/last 10 bp
+  index = which(as.character(seqnames(RCranges)) %in% names(genomeSequence))
+  numAstart = rep(0,nrow(se))  
+  numAstart[index] = letterFrequency(startSeqs, "A")
+  numAend = rep(0,nrow(se)) 
+  numAend[index] = letterFrequency(endSeqs, "A")
+  
+  numTstart = rep(0,nrow(se))  
+  numTstart[index] = letterFrequency(startSeqs, "T")
+  numTend = rep(0,nrow(se)) 
+  numTend[index] = letterFrequency(endSeqs, "T")
+  
+  rowData(se)$numAstart = numAstart
+  rowData(se)$numAend = numAend
+  rowData(se)$numTstart = numTstart
+  rowData(se)$numTend = numTend
+  return(se)
+}
+
+#' calculates a score based on how likely the read class is associated with a 
+#' real gene
+getGeneScore = function(se, thresholdIndex){
+  geneFeatures = prepareGeneModelFeatures(se[thresholdIndex,])
+  if(checkFeatures(geneFeatures)){
+  if(sum(geneFeatures$labels)==length(geneFeatures$labels) | 
+    sum(geneFeatures$labels)==0){geneModel = NULL}
+  geneModel = fit_xgb(geneFeatures$features,geneFeatures$labels)$cvfit
+  geneScore = as.numeric(predict(geneModel, as.matrix(geneFeatures$features), 
+      s = "lambda.min",type="response"))
+  names(geneScore) = geneFeatures$names
+  
+  geneScore = geneScore[order(geneScore, decreasing = T)]  
+  geneFDR = cumsum(geneFeatures$labels)/(1:length(geneScore))
+  geneFDR = rev(cummin(rev(geneFDR)))
+  names(geneFDR) = names(geneScore)
+
+  rowData(se)$geneScore = geneScore[rowData(se)$GENEID]
+  rowData(se)$geneFDR = geneFDR[rowData(se)$GENEID]
+
+  } else {
+    message("Gene Score not calculated")
+    rowData(se)$geneScore = rep(1,nrow(se))
+    rowData(se)$geneFDR = rep(1,nrow(se))
+  }
+  return(se)  
+}
+
+#' calculate and format features by gene for model
+prepareGeneModelFeatures = function(se){
+  temp = by(rowData(se)$readCount, rowData(se)$GENEID, sum)
+  geneIDs = names(temp)
+  labels = !grepl("gene.",geneIDs)
+  numReads = as.numeric(temp)
+  numReadsLog = log(numReads,2)
+  numReadsLog[is.infinite(numReadsLog)]=0
+
+  strand_bias = 1-abs(0.5-(as.numeric(by(rowData(se)$readCount.posStrand, 
+    rowData(se)$GENEID, sum))/numReads))
+  strand_bias[is.na(strand_bias)]=0
+  #how many read classes does a gene have
+  numRCs = table(rowData(se)$GENEID)
+  
+  #number of non-subset read classes
+  subset = rowData(se)$compatible >= 2 | 
+              (rowData(se)$compatible == 1 & 
+              !rowData(se)$equal)
+  subsetRCs = table(rowData(se)$GENEID[subset])
+  temp = numRCs
+  temp[names(subsetRCs)] = temp[names(subsetRCs)]-subsetRCs
+  numNonSubsetRCs = temp
+  #max number of exons of the read classes for each gene
+  numExons = as.numeric(by(rowData(se)$numExons, rowData(se)$GENEID, max))
```

Comment:
6.4sec -> 2.2sec

Reply:
changed in 56f95a1

Reply:
the change in 56f95a1 is actually slower.
3.36 seconds -> 4.4sec
Calculating the matching introns seems to be worse than just including them in the overlap calculation. Will revert 56f95a1

Reply:
latest version is 1b3a112

Reply:
are these 2 commits unrelated? regarding my original comment here, do you actually use the  non subset information? it might be good to add a comment for htis line what it calculates, it's a rather complicated logic!

Reply:
Yes non subset information should be used, fix bug with it and its now included in the features as of b87282c
The subset logic is you are considered a subset if you are compatible with 2 or more read classes (can be equal or non equal), OR you are only compatible with 1 read class but you are not equal. Does this make sense?

Reply:
ok! it might fail in the case of two identical non subset transcripts, which happens but which is rare